File: Client/src/main/java/com/rspsi/editor/model/FloorId.java
package com.rspsi.editor.model;

/**
 * The one conversion between floor ids as stored in map data and floor
 * definition ids.
 *
 * <p>Map tiles (and {@link TileSnapshot#underlayId()}/{@link TileSnapshot#overlayId()})
 * store {@code definitionId + 1}, with 0 meaning "no floor"; the client
 * resolves them with {@code getOverlay(value - 1)}
 * ({@code runescape-client/class470}, the scene tile builder). Tools and
 * painter state keep the encoded value so it round-trips with tiles; UI
 * labels and definition lookups use {@link #definitionId(int)}.</p>
 */
public final class FloorId {
    private FloorId() {
    }

    /** Definition id for an encoded map value, or -1 when the tile has no floor. */
    public static int definitionId(int encoded) {
        return encoded > 0 ? encoded - 1 : -1;
    }

    /** Encoded map value for a definition id, or 0 for "no floor". */
    public static int encode(int definitionId) {
        return definitionId >= 0 ? definitionId + 1 : 0;
    }
}

File: Client/src/main/kotlin/com/rspsi/editor/model/FloorId.kt
@file:JvmName("FloorId")

package com.rspsi.editor.model

/**
 * Converts between floor ids as encoded in map tiles and floor-definition ids.
 *
 * Map tiles store `definitionId + 1`, reserving 0 for "no floor". Keeping this conversion in
 * one neutral boundary prevents tools, inspectors, and cache code from each open-coding the
 * offset and disagreeing about the sentinel.
 */

/** Returns the definition id for an encoded map value, or -1 when the tile has no floor. */
fun definitionId(encoded: Int): Int =
    if (encoded > 0) encoded - 1 else -1

/** Returns the encoded map value for a definition id, or 0 for "no floor". */
fun encode(definitionId: Int): Int =
    if (definitionId >= 0) definitionId + 1 else 0
